segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableEntryIterator.java

```diff
@@ -0,0 +1,328 @@
+            return CompletableFuture.completedFuture(Collections.emptyList());
+        }
+        ReadResult result = this.segment.read(startOffset, length, timer.getRemaining());
+        return AsyncReadResultProcessor.processAll(result, this.executor, timer.getRemaining())
+                .thenApply(data -> parseEntries(data, startOffset, length));
+    }
+
+    @SneakyThrows(IOException.class)
+    private List<Map.Entry<EntryIteratorState, TableEntry>> parseEntries(BufferView data, long startOffset, int readLength) {
+
+        long maxOffset = startOffset + readLength;
+        long currentOffset = startOffset;
+
+        InputStream input = data.getReader();
+        List<Map.Entry<EntryIteratorState, TableEntry>> entries = new ArrayList<>();
+        try {
+            while (currentOffset < maxOffset) {
+                val entry = AsyncTableEntryReader.readEntryComponents(input, currentOffset, this.entrySerializer);
+                boolean reachedEnd = currentOffset + entry.getHeader().getTotalLength() >= maxLength + startOffset;
+                // We must preserve deletions to accurately construct a delta.
+                byte[] value = entry.getValue() == null ? new byte[0] : entry.getValue();
+                entries.add(new AbstractMap.SimpleEntry<>(
+                        new EntryIteratorState(currentOffset, reachedEnd, shouldClear, entry.getHeader().isDeletion()),
+                        TableEntry.versioned(new ByteArraySegment(entry.getKey()), new ByteArraySegment(value), entry.getVersion())));
+                currentOffset += entry.getHeader().getTotalLength();
+            }
+
+        } catch (EOFException ex) {
+            input.close();
+        }
+        this.currentBatchOffset = currentOffset;
+
+        return entries;
+    }
+
+    /**
+     * Creates a new {@link TableIterator} that contains no elements.
+     *
+     * @param <T> Type of elements returned at each iteration.
+     * @return A new instance of the {@link TableIterator.Builder} class.
+     */
+    static <T> TableEntryIterator<T> empty() {
+        return new TableEntryIterator<>(
+                null,
+                0L,
+                0,
+                false,
+                Duration.ofMillis(0),
+                new EntrySerializer(),
+                ignored -> CompletableFuture.completedFuture(null),
+                ForkJoinPool.commonPool());
+    }
+
+    //endregion
+
+    //region Builder
+
+    /**
+     * Creates a new {@link TableIterator.Builder} that can be used to construct {@link TableIterator} instances.
+     *
+     * @param <T> Type of the elements returned at each iteration.
+     * @return A new instance of the {@link TableIterator.Builder} class.
+     */
+    static <T> Builder<T> builder() {
+        return new Builder<>();
+    }
+
+    /**
+     * Builder for the {@link TableIterator} class.
+     */
+    static class Builder<T> {
+
+        private int maxLength = MAX_READ_SIZE;
```

Comment:
You don't need to declare this class. Adding `@Builder` will do it for you.

The reason I did it "manually" in `TableIterator` was because `build()` had to do a significant amount of custom work, so the `@Builder` annotation wouldn't have helped.

In your case, however, this is plain boilerplate code so that annotation will save you from having to do that.